base/set.jl

```diff
@@ -196,6 +221,10 @@ function _unique!(A::AbstractVector)
     resize!(A, count)
 end
 
+# If A is not grouped, then we will need to keep track of all of the elements that we have
+# seen so far.
+_unique!(A::AbstractVector) = unique!(identity, A::AbstractVector)
```

Comment:
Since the call to `unique!` here will now use a `Set{Any}` this will decrease regular `unique!(f,A)`s performance, and I think Andy's objection should be fixed in this PR, since this PR not only adding a new feature that can be optimized later, it makes performance of existing code worse. It would at least be good if you could present some benchmarks.

Reply:
Okay, I will try to build the inner _unique function as suggested by @andyferris.In the meantime I will benchmark the existing solution and will present it here.

Reply:
@fredrikekre When I run @btime unique!(x->2x, 1:10_000); it says setindex! not defined for UnitRange{Int64}. I couldn’t understand what the error message says?
Same problem occurs with the @benchmark macro.

Reply:
It's because `1:10_000` is an immutable array, try instead `@btime unique!(x->2x, $([1:10_000;]))`

Reply:
Some benchmarking results for unique!(f, iter) and unique!(iter) are stated:
```@benchmark unique!(x->x^2, $([1:10_000;]))
BenchmarkTools.Trial: 
  memory estimate:  786.97 KiB
  allocs estimate:  37981
  --------------
  minimum time:     995.317 μs (0.00% GC)
  median time:      1.050 ms (0.00% GC)
  mean time:        1.159 ms (3.08% GC)
  maximum time:     38.007 ms (96.42% GC)
  --------------
  samples:          4305
  evals/sample:     1
```
```@benchmark unique!($([1:10_000;]))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     38.053 μs (0.00% GC)
  median time:      38.081 μs (0.00% GC)
  mean time:        41.537 μs (0.00% GC)
  maximum time:     10.190 ms (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

Reply:
Sorry, I lead you down the wrong path with my coimment on https://github.com/JuliaLang/julia/pull/30141#discussion_r236053015

Replacing `_unique!(iter)` with `unique!(identity, iter)` would only have been optimised away if the `Set{Any}` had not been needed.

Reply:
I have a branch, `ajf/faster_unique_f`. Completely untested, but I'll finish it up.

@fredrikekre I was thinking it would be better to have small PRs with seperate concerns, as Lyndon suggests. Let's simply merge this *feature* PR from @raghav9-97 (his first Julia PR if I understand) and I'll follow up straight away with the *performance* PR that addresses both `unique` and `unique!`. Are you OK with that?

Reply:
Ok.

Reply:
But yeah, this particular line does leave us in an bad intermediate state... oh well.